### PR TITLE
[WIP] Add the ability to reload keys when they are changed

### PR DESCRIPTION
**Pull request by AdamBJohnsonx**

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 1]
```diff
@@ -30,9 +30,11 @@
 import com.microsoft.identity.common.adal.internal.util.StringExtensions;
 import com.microsoft.identity.common.internal.logging.Logger;
 
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:37]:
```suggestion
```

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:37]: unused

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 2]
```diff
@@ -57,7 +59,17 @@
...
     private final Map<String, byte[]> mBrokerSecretKeys = new HashMap<String, byte[]>(2);
 
-    private AtomicReference<byte[]> mSecretKeyData = new AtomicReference<>();
+    private byte[] mSecretKeyData = null;
+
+    /**
+     * @return the version of the key state.  This will be incremented on any alteration, and is
+     * an indication that the state should be reloaded.
+     */
+    public long getSecretKeyVersion() {
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:68]:
```suggestion
    public synchronized long getSecretKeyVersion() {
```

File: common/src/androidTest/java/com/microsoft/identity/common/SharedPreferencesFileManagerTests.java [hunk 3]
```diff
@@ -98,6 +99,17 @@ public void testGetString() {
...
 
+    @Test
+    public void testGetStringWithKeyChange() {
+        mSharedPreferencesFileManager.putString(sTEST_KEY, sTEST_VALUE);
+        assertEquals(sTEST_VALUE, mSharedPreferencesFileManager.getString(sTEST_KEY));
+        AuthenticationSettings.INSTANCE.setSecretKey(null);
+        assertFalse(sTEST_VALUE.equals(mSharedPreferencesFileManager.getString(sTEST_KEY)));
+        mSharedPreferencesFileManager.putString(sTEST_KEY, sTEST_VALUE);
+        assertEquals(sTEST_VALUE, mSharedPreferencesFileManager.getString(sTEST_KEY));
+        AuthenticationSettings.INSTANCE.setSecretKey(new SecureRando);
```
Comment (AdamBJohnsonx) [inline common/src/androidTest/java/com/microsoft/identity/common/SharedPreferencesFileManagerTests.java:110]:
```suggestion
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java [hunk 4]
```diff
@@ -67,6 +68,13 @@
     private static final ConcurrentMap<String, SharedPreferencesFileManager> objectCache =
             new ConcurrentHashMap<String, SharedPreferencesFileManager>(16, 0.75f , 1);
 
+    /**
+     * Clear all cache data from sharedPreferences.
+     */
+    public static void clearAll() {
+        objectCache.clear();
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java:75]: This is a bit confusing - when is one supposed to call this?

File: common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java [hunk 5]
```diff
@@ -151,29 +179,20 @@ public SharedPreferencesFileManager(
...
-            Logger.verbose(TAG, "Init with operating mode: " + TAG);
-        } else if (operatingMode == -1) {
-            Logger.verbose(TAG, "Init with storage helper:  " + TAG);
-        } else {
-            Logger.verbose(TAG, "Init with operating mode and storage helper " + TAG);
-        }
-        mSharedPreferencesFileName = name;
-        mSharedPreferences = context.getSharedPreferences(name, operatingMode == -1 ? Context.MODE_PRIVATE : operatingMode);
-        mStorageHelper = storageHelper;
+        this(context, name, operatingMode, null, new LruCache<String, String>(256));
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java:182]:
```suggestion
        this(context, name, operatingMode, storageHelper, new LruCache<String, String>(256));
```

I assume we don't mean to discard the provided instance?

File: common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java (hunk 4 above)
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java]: Documented.

**Pull request closed without merging**